sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/reflect/DoFnSignaturesTest.java

```diff
@@ -188,6 +193,105 @@ public void foo(ProcessContext context) {}
   }
 
   @Test
+  public void testStateParameterNoAnnotation() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("missing StateId annotation");
+    thrown.expectMessage("myProcessElement");
+    thrown.expectMessage("index 1");
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.getSignature(
+            new DoFn<KV<String, Integer>, Long>() {
+              @ProcessElement
+              public void myProcessElement(
+                  ProcessContext context, ValueState<Integer> noAnnotation) {}
+            }.getClass());
+  }
+
+  @Test
+  public void testStateParameterUndeclared() throws Exception {
+    thrown.expect(IllegalArgumentException.class);
+    thrown.expectMessage("undeclared");
+    thrown.expectMessage("my-state-id");
+    thrown.expectMessage("myProcessElement");
+    thrown.expectMessage("index 1");
+    DoFnSignature sig =
+        DoFnSignatures.INSTANCE.getSignature(
+            new DoFn<KV<String, Integer>, Long>() {
+              @ProcessElement
+              public void myProcessElement(
+                  ProcessContext context, @StateId("my-state-id") ValueState<Integer> undeclared) {}
+            }.getClass());
+  }
+
+  @Test
+  public void testStateParameterDuplicate() throws Exception {
```

Comment:
Consider putting state-related tests in a separate test file - e.g. DoFnSignaturesStateTest (like the ones for SplittableDoFn).